src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,117 @@
+        internal abstract SyntaxNode GetTypeBlock(SyntaxNode syntaxNode);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(context.Span.Start);
+
+            var property = GetProperty(token);
+            if (property == null)
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var propertySymbol = semanticModel.GetDeclaredSymbol(property) as IPropertySymbol;
+            if (propertySymbol == null)
+            {
+                return;
+            }
+
+            if (!(IsValidAutoProperty(property, propertySymbol)))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(
+                new ConvertAutoPropertyToFullPropertyCodeAction(
+                    FeaturesResources.Convert_to_full_property,
+                    c => ExpandToFullPropertyAsync(document, property, propertySymbol, root, cancellationToken)));
+        }
+
+        internal bool IsValidAutoProperty(SyntaxNode property, IPropertySymbol propertySymbol)
+        {
+            var fields = propertySymbol.ContainingType.GetMembers().OfType<IFieldSymbol>();
+            var field = fields.FirstOrDefault(f => propertySymbol.Equals(f.AssociatedSymbol));
+            return field != null;
+        }
+
+        private async Task<Document> ExpandToFullPropertyAsync(
+            Document document,
+            SyntaxNode property,
+            IPropertySymbol propertySymbol,
+            SyntaxNode root,
+            CancellationToken cancellationToken)
+        {
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var workspace = document.Project.Solution.Workspace;
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+
+            // Create full property. If the auto property had an initial value
+            // we need to remove it and later add it to the backing field
+            var fieldName = await GetFieldNameAsync(document, propertySymbol, cancellationToken).ConfigureAwait(false);
+            var accessorTuple = GetNewAccessors(options, property, fieldName, generator);
+            var fullProperty = generator
+                .WithAccessorDeclarations(
+                    GetPropertyWithoutInitializer(property),
+                    accessorTuple.newSetAccessor == null
+                        ? new SyntaxNode[] { accessorTuple.newGetAccessor }
+                        : new SyntaxNode[] { accessorTuple.newGetAccessor, accessorTuple.newSetAccessor })
```

Comment:
i think it *might* be possible to just use "new [] { }" instead of "new SyntaxNode[] { }"